### ISSUE TITLE
fit_generator slows down when dealing with large dataset

### ISSUE THREAD

**Issue by yushuinanrong** [fchollet/keras#5390]
Hi all,

I'm working on a image SR problem and my dataset has ~300K  32x32 pixel images. I'm using fit_generator to fetch images with a batch size of 128 in training. 
I just find when I only use a small portion of the dataset (~30K), the training for every batch is fairly fast. However, if I use the whole dataset instead, the training for every batch becomes much slower (~10 times slower).
I'm confused because fit_generator only fetch a batch of images every iteration, it should not be sensitive to the number of overall training samples....

I'm using tensorflow backend on Ubuntu 16.04 with a Titan X GPU.

Any idea will be greatly appreciated.

**Comment by unrealwill** [comment]
Use SSD drive, or prefetch images. You probably are hitting disk. You can monitor gpu usage using nvidia-smi, and probably you can't fill your nice GPU fast enough. Sometimes python also lags because of garbage collection, so bigger RAM could help.
You could probably benefit for using more complex models, so that you have a better gpu usage.
I think Keras doesn't load data asynchronously.

**Comment by yushuinanrong** [comment]
@unrealwill 
Thanks for your answer. Yes, I store my data in disk. It could be the reason.
I also find the training only slows down in the first epoch, the later epochs become faster as I expected.

**Comment by unrealwill** [comment]
Well it's probably that the OS, is transparently caching files in RAM. You will hit this barrier again as dataset size grow beyond RAM size, but you could probably asynchronously (from another thread) prefetch the next files so that OS cache them and you hit a hot cache.

**Comment by yushuinanrong** [comment]
@unrealwill 
I will try your suggestion and update later.
Thanks again.

**Comment by ppwwyyxx** [comment]
@yushuinanrong [Here](http://tensorpack.readthedocs.io/en/latest/tutorial/efficient-dataflow.html) is a tutorial about how to build a faster generator to load data and you can use it in Keras as well. It should be fairly easy to handle your dataset size.

**Comment by yushuinanrong** [comment]
@ppwwyyxx 
Thanks for the reference. I just note your comment, and even months later, I feel like the tutorial is quite useful for me.

**Comment by jsl303** [comment]
Hi,
I also have the slow performance issue with fit_generator.
Even if I process everything on memory, it's very slow.
For test, I initialized generators with 10000x180 features and 10000x20 multi labels data.
Then getitem simply slices features and labels according to the index and return the batch. There's no process between.
However, the fit_generator and predict_generator both takes more than several times longer than just fit and predict.
If I google slow Keras fit_generator, I can find many other posts describing the performance issue.
